blockchain/chain.go

```diff
@@ -1254,6 +1254,109 @@ func (b *BlockChain) BestSnapshot() *BestState {
 	return snapshot
 }
 
+// TipStatus is the status of a chain tip.
+type TipStatus byte
+
+const (
+	// StatusUnknown indicates that the tip status isn't any of the defined
+	// statuses.
+	StatusUnknown TipStatus = iota
+	// StatusActive indicates that the tip is considered active and is in
+	// the best chain.
+	StatusActive
+	// StatusInvalid indicates that this tip or any of the ancestors of this
+	// tip are invalid.
+	StatusInvalid
+	// StatusValidFork is given if:
+	// 1: Not a part of the best chain.
+	// 2: Is not invalid.
+	// 3: Has the block data stored to disk.
+	StatusValidFork
+)
+
+// String returns the status flags as string.
+func (ts TipStatus) String() string {
+	switch ts {
+	case StatusActive:
+		return "active"
+	case StatusInvalid:
+		return "invalid"
+	case StatusValidFork:
+		return "valid-fork"
+	}
+	return "unknown"
+}
+
+// ChainTip represents the last block in a branch of the block tree.
+type ChainTip struct {
+	// Height of the tip.
+	Height int32
+	// BlockHash hash of the tip.
+	BlockHash chainhash.Hash
+	// Amount of blocks connecting this tip with the best chain. Returns
```

Comment:
A bit confused re the wording here: this is meant to the the height gap between this non-main chain tip and the main chain itself?

Reply:
It's meant as the amount of blocks the branch has from the oldest common ancestor with the active chain. Active chain always returns 0.

The below branch of "unknown" would have a `branchlen` of 2 since it forks off at block 10 and has 2 blocks built on top of block 10.
```
//      genesis -> 1 -> 2 -> 3 ... -> 10 -> 11  -> 12  -> 13 (active)
//                                      \-> 11a -> 12a (unknown)
```

Reply:
Makes a lot of sense, I think we can add some more of this context to the comment itself (can even add that diagram above).